File: vscode-vim-mentor/src/feedback.ts
// =============================================================================
// feedback.ts
// Success feedback system: flash decorations, streak tracking, milestones.
// =============================================================================

import * as vscode from "vscode";
import * as storage from "./storage";

// -----------------------------------------------------------------------------
// Decoration Types
// -----------------------------------------------------------------------------

const successFlashDecoration = vscode.window.createTextEditorDecorationType({
  backgroundColor: new vscode.ThemeColor(
    "editorInfo.background"
  ),
  isWholeLine: true,
});

// Fallback with explicit color if theme color not available
const successFlashDecorationFallback =
  vscode.window.createTextEditorDecorationType({
    backgroundColor: "rgba(78, 154, 6, 0.15)",
    isWholeLine: true,
  });

// -----------------------------------------------------------------------------
// State
// -----------------------------------------------------------------------------

interface FeedbackState {
  streak: number;
  totalCorrect: number;
  totalIncorrect: number;
  sessionStart: number;
}

const state: FeedbackState = {
  streak: 0,
  totalCorrect: 0,
  totalIncorrect: 0,
  sessionStart: Date.now(),
};

const MILESTONES = [5, 10, 25, 50, 100];

// -----------------------------------------------------------------------------
// Flash Effect
// -----------------------------------------------------------------------------

/**
 * Flash the current line green briefly to indicate success.
 */
export function flashSuccess(durationMs: number = 300): void {
  const editor = vscode.window.activeTextEditor;
  if (!editor) {
    return;
  }

  const line = editor.selection.active.line;
  const range = new vscode.Range(line, 0, line, 0);

  editor.setDecorations(successFlashDecoration, [range]);

  setTimeout(() => {
    editor.setDecorations(successFlashDecoration, []);
  }, durationMs);
}

// -----------------------------------------------------------------------------
// Streak and Milestone Tracking
// -----------------------------------------------------------------------------

/**
 * Record a correct Vim command usage. Increments streak and triggers feedback.
 */
export function recordSuccess(command: string): void {
  state.streak++;
  state.totalCorrect++;

  // Flash
  flashSuccess();

  // Persist
  storage.recordCorrect(command);
  storage.updateBestStreak(state.streak);

  // Check milestones
  for (const milestone of MILESTONES) {
    if (state.streak === milestone) {
      showMilestone(milestone);
      break;
    }
  }
}

/**
 * Record a failed (non-Vim) operation. Resets streak.
 */
export function recordFailure(command: string): void {
  state.streak = 0;
  state.totalIncorrect++;

  storage.recordPrompted(command);
}

/**
 * Show a milestone celebration notification.
 */
function showMilestone(count: number): void {
  const messages: Record<number, { title: string; body: string }> = {
    5: {
      title: "Nice start!",
      body: "5 correct Vim commands in a row!",
    },
    10: {
      title: "Getting better!",
      body: "10 streak! You are learning fast.",
    },
    25: {
      title: "Impressive!",
      body: "25 streak! Vim is becoming natural.",
    },
    50: {
      title: "Vim Apprentice!",
      body: "50 streak! True muscle memory forming.",
    },
    100: {
      title: "Vim Master!",
      body: "100 streak! Nothing can stop you now.",
    },
  };

  const msg = messages[count] ?? {
    title: "Milestone!",
    body: `${count} correct commands in a row!`,
  };

  vscode.window.showInformationMessage(
    `Vim Mentor: ${msg.title} - ${msg.body}`
  );
}

// -----------------------------------------------------------------------------
// State Accessors
// -----------------------------------------------------------------------------

export function getStreak(): number {
  return state.streak;
}

export function getStats(): {
  streak: number;
  totalCorrect: number;
  totalIncorrect: number;
  accuracy: number;
} {
  const total = state.totalCorrect + state.totalIncorrect;
  return {
    streak: state.streak,
    totalCorrect: state.totalCorrect,
    totalIncorrect: state.totalIncorrect,
    accuracy: total > 0 ? Math.floor((state.totalCorrect / total) * 100) : 0,
  };
}

export function getSessionStart(): number {
  return state.sessionStart;
}

export function resetSession(): void {
  state.streak = 0;
  state.totalCorrect = 0;
  state.totalIncorrect = 0;
  state.sessionStart = Date.now();
}

// -----------------------------------------------------------------------------
// Cleanup
// -----------------------------------------------------------------------------

export function dispose(): void {
  successFlashDecoration.dispose();
  successFlashDecorationFallback.dispose();
}

File: lua/vim-mentor/ui/feedback.lua
-- vim-mentor/ui/feedback.lua
-- Positive reinforcement and success feedback system
-- Provides visual/auditory feedback when users correctly use Vim commands

local M = {}

local config = {}

local state = {
  streak = 0,
  total_correct = 0,
  total_incorrect = 0,
  session_start = nil,
}

local flash_ns = vim.api.nvim_create_namespace("vim_mentor_flash")
local streak_ns = vim.api.nvim_create_namespace("vim_mentor_streak")

-- ============================================================
-- Success Flash Effect
-- Briefly highlights the cursor line green on correct input
-- ============================================================

--- Flash the cursor line to indicate success
---@param duration number|nil Duration in ms (default from config)
function M.flash_success(duration)
  if not config.feedback.enabled or not config.feedback.success_flash then
    return
  end

  duration = duration or config.feedback.success_flash_duration

  local line = vim.api.nvim_win_get_cursor(0)[1] - 1
  vim.api.nvim_buf_set_extmark(0, flash_ns, line, 0, {
    line_hl_group = "VimMentorSuccess",
    end_line = line + 1,
    priority = 200,
  })

  vim.defer_fn(function()
    vim.api.nvim_buf_clear_namespace(0, flash_ns, 0, -1)
  end, duration)
end

-- ============================================================
-- Streak Management and Milestone Feedback
-- ============================================================

--- Record a correct Vim operation
function M.record_success()
  state.streak = state.streak + 1
  state.total_correct = state.total_correct + 1

  -- Flash feedback
  M.flash_success()

  -- Check milestones
  if config.feedback.streak.enabled then
    for _, milestone in ipairs(config.feedback.streak.milestones) do
      if state.streak == milestone then
        M.show_milestone(milestone)
        break
      end
    end
  end
end

--- Record a failed (non-Vim) operation
function M.record_failure()
  state.streak = 0
  state.total_incorrect = state.total_incorrect + 1
end

--- Show a milestone celebration popup
---@param count number The milestone number reached
function M.show_milestone(count)
  local messages = {
    [5]   = { title = "Nice start!",      body = "5 correct Vim commands in a row!" },
    [10]  = { title = "Getting better!",   body = "10 streak! You are learning fast." },
    [25]  = { title = "Impressive!",       body = "25 streak! Vim is becoming natural." },
    [50]  = { title = "Vim Apprentice!",   body = "50 streak! True muscle memory." },
    [100] = { title = "Vim Master!",       body = "100 streak! Nothing can stop you." },
  }

  local msg = messages[count] or {
    title = "Milestone!",
    body = string.format("%d correct commands in a row!", count),
  }

  -- Use vim.notify to show milestone (integrates with nvim-notify if available)
  vim.notify(
    string.format("%s\n%s", msg.title, msg.body),
    vim.log.levels.INFO,
    {
      title = "Vim Mentor",
      icon = config.statusline.icons.streak,
      timeout = 3000,
    }
  )
end

-- ============================================================
-- Extended Help (Escalation)
-- Shown after repeated failures on the same operation
-- ============================================================

--- Show extended help for a Vim command
---@param hint table The hint data with command details
function M.show_extended_help(hint)
  local buf = vim.api.nvim_create_buf(false, true)

  local lines = {
    " Extended Help ",
    string.rep("═", 50),
    "",
    string.format(" You have tried this %d times.", hint.attempt_count or 0),
    "",
    " Command:  " .. (hint.keys or "?"),
    " Action:   " .. (hint.desc or ""),
    "",
    " How it works:",
    "   " .. (hint.extended_desc or "No extended description available."),
    "",
    " Related commands:",
  }

  -- Add related commands if available
  if hint.related then
    for _, related in ipairs(hint.related) do
      table.insert(lines, string.format("   %s  -  %s", related.keys, related.desc))
    end
  else
    table.insert(lines, "   (none)")
  end

  table.insert(lines, "")
  table.insert(lines, " Press <Esc> or q to close.")

  vim.api.nvim_buf_set_lines(buf, 0, -1, false, lines)
  vim.api.nvim_buf_set_option(buf, "modifiable", false)
  vim.api.nvim_buf_set_option(buf, "bufhidden", "wipe")

  local width = 55
  local height = #lines
  local win = vim.api.nvim_open_win(buf, true, {
    relative = "editor",
    width = width,
    height = height,
    row = math.floor((vim.o.lines - height) / 2),
    col = math.floor((vim.o.columns - width) / 2),
    border = "rounded",
    title = " Vim Mentor - Extended Help ",
    title_pos = "center",
    style = "minimal",
  })

  -- Highlight setup
  local ns = vim.api.nvim_create_namespace("vim_mentor_extended_help")
  vim.api.nvim_buf_add_highlight(buf, ns, "VimMentorTitle", 0, 0, -1)
  vim.api.nvim_buf_add_highlight(buf, ns, "VimMentorBorder", 1, 0, -1)
  vim.api.nvim_buf_add_highlight(buf, ns, "VimMentorDetected", 3, 0, -1)
  vim.api.nvim_buf_add_highlight(buf, ns, "VimMentorKey", 5, 12, -1)
  vim.api.nvim_buf_add_highlight(buf, ns, "VimMentorDesc", 6, 12, -1)

  -- Close mappings
  local close = function()
    if vim.api.nvim_win_is_valid(win) then
      vim.api.nvim_win_close(win, true)
    end
  end
  vim.keymap.set("n", "<Esc>", close, { buffer = buf, nowait = true })
  vim.keymap.set("n", "q", close, { buffer = buf, nowait = true })
end

-- ============================================================
-- State Accessors (for statusline and progress)
-- ============================================================

function M.get_streak()
  return state.streak
end

function M.get_stats()
  return {
    streak = state.streak,
    total_correct = state.total_correct,
    total_incorrect = state.total_incorrect,
    accuracy = state.total_correct + state.total_incorrect > 0
      and math.floor(state.total_correct / (state.total_correct + state.total_incorrect) * 100)
      or 0,
  }
end

function M.reset_session()
  state.streak = 0
  state.total_correct = 0
  state.total_incorrect = 0
  state.session_start = os.time()
end

-- ============================================================
-- Setup
-- ============================================================

function M.setup(opts)
  config = opts
  state.session_start = os.time()
end

return M
